numpy/lib/function_base.py

```diff
@@ -3196,8 +3189,14 @@ def i0(x):
 
     """
     x = np.asanyarray(x)
-    x = np.abs(x)
-    return piecewise(x, [x <= 8.0], [_i0_1, _i0_2])
+    ind1 = (np.abs(x) <= 8.0)
+    y1 = _i0_1(x[ind1])
+    ind2 = ~ind1
+    y2 = _i0_2(x[ind2])
+    y = np.empty_like(x, dtype=np.common_type(y1, y2))
+    y[ind1] = y1
+    y[ind2] = y2
+    return y
```

Comment:
Am I missing something? I can't see why one of these would fail but the other would work.

Reply:
From the documentation of `piecewise`:
>The output is the same shape and type as x

x = abs(x) is always real, therefore the output of `i0` will always be real.

Reply:
And of course `abs(x)` discards the phase completely.

Reply:
Oh, that's irritating.